File: 234.Palindrome-Linked-List.go
// https://leetcode.com/problems/palindrome-linked-list/
//
// algorithms
// Easy (34.81%)
// Total Accepted:    214,702
// Total Submissions: 616,829
// beats 100.0% of golang submissions

package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	oneStep, twoStep := head, head
	var pre, tmp, preTmp *ListNode
	var left, right *ListNode

	for {
		if twoStep.Next == nil {
			left = pre
			right = oneStep.Next
			break
		}
		if twoStep.Next.Next == nil {
			left = oneStep
			right = oneStep.Next
			left.Next = pre
			break
		}
		twoStep = twoStep.Next.Next
		tmp = oneStep.Next
		preTmp = pre
		pre = oneStep
		pre.Next = preTmp

		oneStep = tmp
	}

	for left != nil && right != nil {
		if left.Val != right.Val {
			return false
		}
		left = left.Next
		right = right.Next
	}

	return true
}

File: 234.Palindrome-Linked-List.py
# https://leetcode.com/problems/palindrome-linked-list/
#
# algorithms
# Easy (35.29%)
# Total Accepted:    231,671
# Total Submissions: 656,459
# beats 100.0% of python submissions


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head or not head.next:
            return True

        head_tmp = ListNode('#')
        head_tmp.next = head

        one, two, pre = head_tmp, head_tmp, None

        while two and two.next:
            two = two.next.next
            tmp = one.next
            one.next = pre
            pre = one
            one = tmp

        if two:
            l, r = one, one.next
            one.next = pre
        else:
            l, r = pre, one.next

        while r:
            if l.val != r.val:
                return False
            l, r = l.next, r.next

        return True
